src/common/range.rs

```diff
@@ -141,278 +146,454 @@ impl ByteRangeSpec {
+            (first, Some(last)) => {
+                if first > last {
+                    return Err(err());
+                }
+                self.inner.push(ByteRangeSpec::FromTo(first, last));
+            }
+            (first, None) => {
+                self.inner.push(ByteRangeSpec::AllFrom(first));
+            }
+        };
+        Ok(self)
+    }
+
+    /// Add a suffix to the header. Fails if the suffix length is zero.
+    pub fn suffix(mut self, len: u64) -> Result<Self, InvalidRange> {
+        let len = match NonZeroU64::new(len) {
+            Some(len) => len,
+            None => return Err(err()),
+        };
+
+        self.inner.push(ByteRangeSpec::Last(len));
+        Ok(self)
+    }
+
+    /// Finish up the header. Fails if no ranges or suffixes were added.
+    pub fn finish(self) -> Result<Range, InvalidRange> {
+        if self.inner.is_empty() {
+            return Err(err());
+        }
+        Ok(Range::Bytes(self.inner))
+    }
+}
 
+impl std::fmt::Display for Range {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Range::Bytes(ranges) => {
+                write!(f, "bytes=")?;
                 for (i, range) in ranges.iter().enumerate() {
                     if i != 0 {
-                        try!(f.write_str(","));
+                        f.write_str(",")?;
                     }
-                    try!(Display::fmt(range, f));
+                    std::fmt::Display::fmt(range, f)?;
                 }
                 Ok(())
-            },
-            Range::Unregistered(ref unit, ref range_str) => {
-                write!(f, "{}={}", unit, range_str)
-            },
+            }
+            Range::Unregistered { unit, set } => write!(f, "{}={}", unit, set),
         }
     }
 }
 
 impl FromStr for Range {
-    type Err = ::Error;
+    type Err = InvalidRange;
 
-    fn from_str(s: &str) -> ::Result<Range> {
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
         let mut iter = s.splitn(2, '=');
 
         match (iter.next(), iter.next()) {
-            (Some("bytes"), Some(ranges)) => {
-                let ranges = from_comma_delimited(ranges);
-                if ranges.is_empty() {
-                    return Err(::Error::Header);
+            (Some("bytes"), Some(set)) => {
+                let mut invalid = false;
+                let specs: Vec<ByteRangeSpec> = set
+                    .split(',')
+                    .filter_map(|spec: &str| -> Option<ByteRangeSpec> {
+                        let spec = spec.parse();
+                        if spec.is_err() {
+                            invalid = true;
+                        }
+                        spec.ok()
+                    })
+                    .collect();
+                if invalid || specs.is_empty() {
+                    return Err(err());
                 }
-                Ok(Range::Bytes(ranges))
+                Ok(Self::Bytes(specs))
```

Comment:
Might be able to eliminate this `invalid` flag by doing something like

```rust
let specs: Result<Vec<ByteRangeSpec>, _> = set
    .split(',')
    .map(|spec: &str| spec.parse())
    .collect();
match specs {
    Ok(specs) if !specs.is_empty() => Ok(Self::Bytes(specs)),
    _ => Err(err()) // or use the underlying error in the failed parse case?
}
```

Reply:
collecting straight into a result, again something I wasn't aware of. Cool!